### Lists and functions are the most fundamental concepts for functional languages.
### Did you know that Lisp stands for List Processor?

In [2]:
let rec len (list): int =
  match list with
  | [] -> 0
  | [_] -> 1
  | _ :: list' -> 1 + (len list')

val len : 'a list -> int = <fun>


In [40]:
let length (list): int =
  match list with
  | [] -> 0
  | [_] -> 1
  | _ :: list' -> 1 + (length list')

error: compile_error

In [3]:
len([1; 2 ;3])

- : int = 3


In [7]:
len[(1,2); (2,3); (4,5); (1,3)]

- : int = 4


In [8]:
len([])

- : int = 0


In [10]:
len(2::[3])

- : int = 2


### Can you compute the sum of a list now?

### Duplicate every element in the list

In [15]:
let rec double (list) =
  match list with
  | [] -> []
  | [x] -> [x; x]
  | x :: list' ->  x :: x :: double list'

val double : 'a list -> 'a list = <fun>


In [16]:
double([1; 2; 3])

- : int list = [1; 1; 2; 2; 3; 3]


### Take a number, create a list of that size

In [18]:
let rec listgen (n: int) =
    if n = 0 then [] else n :: listgen (n-1)

val listgen : int -> int list = <fun>


In [19]:
listgen 1

- : int list = [1]


In [20]:
listgen 10

- : int list = [10; 9; 8; 7; 6; 5; 4; 3; 2; 1]


## CodeWars: Powers of 2: 
https://www.codewars.com/kata/57a083a57cb1f31db7000028/ocaml

In [36]:
open Float

let rec powersOfTwo n = 
  if n = 0 then [1]
  else
     (powersOfTwo (n-1))  @ [to_int(2.** (of_int n)) ]

val powersOfTwo : int -> int list = <fun>


In [37]:
powersOfTwo(2)

- : int list = [1; 2; 4]


In [38]:
powersOfTwo(10)

- : int list = [1; 2; 4; 8; 16; 32; 64; 128; 256; 512; 1024]


In [39]:
powersOfTwo(20)

- : int list =
[1; 2; 4; 8; 16; 32; 64; 128; 256; 512; 1024; 2048; 4096; 8192; 16384; 32768;
 65536; 131072; 262144; 524288; 1048576]


### Tail recursion

https://cs3110.github.io/textbook/chapters/correctness/proving_correctness.html?highlight=tail%20recursion

https://www.geeksforgeeks.org/tail-recursion/

In [41]:
let rec fac n = 
  if n <= 1 then 1
  else n * fac (n-1)

val fac : int -> int = <fun>


In [43]:
fac(5)

- : int = 120


In [44]:
fac(50)

- : int = -3258495067890909184


In [45]:
fac(500)

- : int = 0


In [46]:
fac(5000)

- : int = 0


In [51]:
fac(500000)

error: runtime_error

In [3]:
let fact : int -> int = fun num ->
     let rec factTR : int -> int -> int = fun n acc ->
          print_endline( (string_of_int n) ^ " " ^ (string_of_int acc));
          if n = 1
          then acc
          else factTR (n-1) (acc * n)
     in
          factTR num 1

val fact : int -> int = <fun>


acc is accumulated factorial value

In [5]:
fact(5)

5 1
4 5
3 20
2 60
1 120


- : int = 120


Note that in this case all five factTR return the same value 120.

In [6]:
fact(50)

50 1
49 50
48 2450
47 117600
46 5527200
45 254251200
44 11441304000
43 503417376000
42 21646947168000
41 909171781056000
40 37276043023296000
39 1491041720931840000
38 2810394895213105152
37 -3885458424159313920
36 3811990895781797888
35 -1118908304676913152
34 -2268302516272857088
33 -3335309258438934528
32 615258913772470272
31 1241541167009497088
30 1594288029875306496
29 1711780711985315840
28 3524780463300280320
27 -2763239432994684928
26 -820488396018286592
25 -2885954222765899776
24 1638120725690712064
23 2421409269157986304
22 352180969505030144
21 -1475390707744112640
20 -3313088752062038016
19 -1698170783257329664
18 -4595128771324936192
17 298030447844130816
16 -4156854423504551936
15 -1946066518089400320
14 -1520881660776677376
13 -2845599177163931648
12 -99301155712008192
11 -1191613868544098304
10 -3884380517130305536
9 -1950317023883952128
8 893890858753982464
7 -2072245166822916096
6 3941027905949138944
5 -4023948674869493760
4 -1672999300637917184
3 2531374834303107072

- : int = -3258495067890909184


In [69]:
fact(5000)

5000 1
4999 5000
4998 24995000
4997 124925010000
4996 624250274970000
4995 3118754373750120000
4994 -97273365866939712
4993 3055528813806196096
4992 798018376537920896
4991 -788984243962036224
4990 559498122466476032
4989 -2786096059281670144
4988 -211580216105205760
4987 -3897705731321888768
4986 -4213600449246658560
4985 1829660011339710464
4984 -1059787920916611072
4983 3009179269396955136
4982 -2462632520838021120
4981 -1750409798169395200
4980 -2704629853994352640
4979 -2933499083903467520
4978 3929367622600097792
4977 -2380064865692680192
4976 -2773141230937178112
4975 -986198008653676544
4974 498830554699923456
4973 96101163484577792
4972 -1704259907642982400
4971 2698641068630474752
4970 4161810575245967360
4969 -3824919692804358144
4968 3343814412837322752
4967 776964600368201728
4966 3813658623561695232
4965 3045932944523788288
4964 -3273070881223475200
4963 4057674544784080896
4962 3617609309417897984
4961 1895409612216074240
4960 4510980648927756288
4959 -143654272801492172

IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



612 0
611 0
610 0
609 0
608 0
607 0
606 0
605 0
604 0
603 0
602 0
601 0
600 0
599 0
598 0
597 0
596 0
595 0
594 0
593 0
592 0
591 0
590 0
589 0
588 0
587 0
586 0
585 0
584 0
583 0
582 0
581 0
580 0
579 0
578 0
577 0
576 0
575 0
574 0
573 0
572 0
571 0
570 0
569 0
568 0
567 0
566 0
565 0
564 0
563 0
562 0
561 0
560 0
559 0
558 0
557 0
556 0
555 0
554 0
553 0
552 0
551 0
550 0
549 0
548 0
547 0
546 0
545 0
544 0
543 0
542 0
541 0
540 0
539 0
538 0
537 0
536 0
535 0
534 0
533 0
532 0
531 0
530 0
529 0
528 0
527 0
526 0
525 0
524 0
523 0
522 0
521 0
520 0
519 0
518 0
517 0
516 0
515 0
514 0
513 0
512 0
511 0
510 0
509 0
508 0
507 0
506 0
505 0
504 0
503 0
502 0
501 0
500 0
499 0
498 0
497 0
496 0
495 0
494 0
493 0
492 0
491 0
490 0
489 0
488 0
487 0
486 0
485 0
484 0
483 0
482 0
481 0
480 0
479 0
478 0
477 0
476 0
475 0
474 0
473 0
472 0
471 0
470 0
469 0
468 0
467 0
466 0
465 0
464 0
463 0
462 0
461 0
460 0
459 0
458 0
457 0
456 0
455 0
454 0
453 0
452 0
451 0
450 0
449 0
448 0
447 0
446 

- : int = 0


In [ ]:
fact(5000000)

5000000 1
4999999 5000000
4999998 24999995000000
4999997 -4127283515956861312
4999996 4290813264122506368
4999995 -3039082686482551296
4999994 -3486745523180873216
4999993 -2466017582824856576
4999992 564655329982628864
4999991 -2047810742415859712
4999990 -3962973466806837248
4999989 696892628630421504
4999988 3872390011106738176
4999987 -1907096798582079488
4999986 4075813814984048640
4999985 -831787683160129536
4999984 3192896860391079936
4999983 2928290918184255488
4999982 2677743512675418112
4999981 4068551396130029568
4999980 -2304687724252102656
4999979 2570807186158518272
4999978 -4474624312887738368
4999977 -253369620889600000
4999976 -2904312679227195392
4999975 4603121368039424000
4999974 -3533142461524213760
4999973 -3750008748517621760
4999972 1490325179836923904
4999971 3454760537001295872
4999970 -3897739418051543040
4999969 -872570653981540352
4999968 3997283580865675264
4999967 -896103488465928192
4999966 -4319711392866238464
4999965 -2901952858939195392
4999964 198308

IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



4995157 0
4995156 0
4995155 0
4995154 0
4995153 0
4995152 0
4995151 0
4995150 0
4995149 0
4995148 0
4995147 0
4995146 0
4995145 0
4995144 0
4995143 0
4995142 0
4995141 0
4995140 0
4995139 0
4995138 0
4995137 0
4995136 0
4995135 0
4995134 0
4995133 0
4995132 0
4995131 0
4995130 0
4995129 0
4995128 0
4995127 0
4995126 0
4995125 0
4995124 0
4995123 0
4995122 0
4995121 0
4995120 0
4995119 0
4995118 0
4995117 0
4995116 0
4995115 0
4995114 0
4995113 0
4995112 0
4995111 0
4995110 0
4995109 0
4995108 0
4995107 0
4995106 0
4995105 0
4995104 0
4995103 0
4995102 0
4995101 0
4995100 0
4995099 0
4995098 0
4995097 0
4995096 0
4995095 0
4995094 0
4995093 0
4995092 0
4995091 0
4995090 0
4995089 0
4995088 0
4995087 0
4995086 0
4995085 0
4995084 0
4995083 0
4995082 0
4995081 0
4995080 0
4995079 0
4995078 0
4995077 0
4995076 0
4995075 0
4995074 0
4995073 0
4995072 0
4995071 0
4995070 0
4995069 0
4995068 0
4995067 0
4995066 0
4995065 0
4995064 0
4995063 0
4995062 0
4995061 0
4995060 0
4995059 0
4995058 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



4991104 0
4991103 0
4991102 0
4991101 0
4991100 0
4991099 0
4991098 0
4991097 0
4991096 0
4991095 0
4991094 0
4991093 0
4991092 0
4991091 0
4991090 0
4991089 0
4991088 0
4991087 0
4991086 0
4991085 0
4991084 0
4991083 0
4991082 0
4991081 0
4991080 0
4991079 0
4991078 0
4991077 0
4991076 0
4991075 0
4991074 0
4991073 0
4991072 0
4991071 0
4991070 0
4991069 0
4991068 0
4991067 0
4991066 0
4991065 0
4991064 0
4991063 0
4991062 0
4991061 0
4991060 0
4991059 0
4991058 0
4991057 0
4991056 0
4991055 0
4991054 0
4991053 0
4991052 0
4991051 0
4991050 0
4991049 0
4991048 0
4991047 0
4991046 0
4991045 0
4991044 0
4991043 0
4991042 0
4991041 0
4991040 0
4991039 0
4991038 0
4991037 0
4991036 0
4991035 0
4991034 0
4991033 0
4991032 0
4991031 0
4991030 0
4991029 0
4991028 0
4991027 0
4991026 0
4991025 0
4991024 0
4991023 0
4991022 0
4991021 0
4991020 0
4991019 0
4991018 0
4991017 0
4991016 0
4991015 0
4991014 0
4991013 0
4991012 0
4991011 0
4991010 0
4991009 0
4991008 0
4991007 0
4991006 0
4991005 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



4986988 0
4986987 0
4986986 0
4986985 0
4986984 0
4986983 0
4986982 0
4986981 0
4986980 0
4986979 0
4986978 0
4986977 0
4986976 0
4986975 0
4986974 0
4986973 0
4986972 0
4986971 0
4986970 0
4986969 0
4986968 0
4986967 0
4986966 0
4986965 0
4986964 0
4986963 0
4986962 0
4986961 0
4986960 0
4986959 0
4986958 0
4986957 0
4986956 0
4986955 0
4986954 0
4986953 0
4986952 0
4986951 0
4986950 0
4986949 0
4986948 0
4986947 0
4986946 0
4986945 0
4986944 0
4986943 0
4986942 0
4986941 0
4986940 0
4986939 0
4986938 0
4986937 0
4986936 0
4986935 0
4986934 0
4986933 0
4986932 0
4986931 0
4986930 0
4986929 0
4986928 0
4986927 0
4986926 0
4986925 0
4986924 0
4986923 0
4986922 0
4986921 0
4986920 0
4986919 0
4986918 0
4986917 0
4986916 0
4986915 0
4986914 0
4986913 0
4986912 0
4986911 0
4986910 0
4986909 0
4986908 0
4986907 0
4986906 0
4986905 0
4986904 0
4986903 0
4986902 0
4986901 0
4986900 0
4986899 0
4986898 0
4986897 0
4986896 0
4986895 0
4986894 0
4986893 0
4986892 0
4986891 0
4986890 0
4986889 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



4982948 0
4982947 0
4982946 0
4982945 0
4982944 0
4982943 0
4982942 0
4982941 0
4982940 0
4982939 0
4982938 0
4982937 0
4982936 0
4982935 0
4982934 0
4982933 0
4982932 0
4982931 0
4982930 0
4982929 0
4982928 0
4982927 0
4982926 0
4982925 0
4982924 0
4982923 0
4982922 0
4982921 0
4982920 0
4982919 0
4982918 0
4982917 0
4982916 0
4982915 0
4982914 0
4982913 0
4982912 0
4982911 0
4982910 0
4982909 0
4982908 0
4982907 0
4982906 0
4982905 0
4982904 0
4982903 0
4982902 0
4982901 0
4982900 0
4982899 0
4982898 0
4982897 0
4982896 0
4982895 0
4982894 0
4982893 0
4982892 0
4982891 0
4982890 0
4982889 0
4982888 0
4982887 0
4982886 0
4982885 0
4982884 0
4982883 0
4982882 0
4982881 0
4982880 0
4982879 0
4982878 0
4982877 0
4982876 0
4982875 0
4982874 0
4982873 0
4982872 0
4982871 0
4982870 0
4982869 0
4982868 0
4982867 0
4982866 0
4982865 0
4982864 0
4982863 0
4982862 0
4982861 0
4982860 0
4982859 0
4982858 0
4982857 0
4982856 0
4982855 0
4982854 0
4982853 0
4982852 0
4982851 0
4982850 0
4982849 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



4915463 0
4915462 0
4915461 0
4915460 0
4915459 0
4915458 0
4915457 0
4915456 0
4915455 0
4915454 0
4915453 0
4915452 0
4915451 0
4915450 0
4915449 0
4915448 0
4915447 0
4915446 0
4915445 0
4915444 0
4915443 0
4915442 0
4915441 0
4915440 0
4915439 0
4915438 0
4915437 0
4915436 0
4915435 0
4915434 0
4915433 0
4915432 0
4915431 0
4915430 0
4915429 0
4915428 0
4915427 0
4915426 0
4915425 0
4915424 0
4915423 0
4915422 0
4915421 0
4915420 0
4915419 0
4915418 0
4915417 0
4915416 0
4915415 0
4915414 0
4915413 0
4915412 0
4915411 0
4915410 0
4915409 0
4915408 0
4915407 0
4915406 0
4915405 0
4915404 0
4915403 0
4915402 0
4915401 0
4915400 0
4915399 0
4915398 0
4915397 0
4915396 0
4915395 0
4915394 0
4915393 0
4915392 0
4915391 0
4915390 0
4915389 0
4915388 0
4915387 0
4915386 0
4915385 0
4915384 0
4915383 0
4915382 0
4915381 0
4915380 0
4915379 0
4915378 0
4915377 0
4915376 0
4915375 0
4915374 0
4915373 0
4915372 0
4915371 0
4915370 0
4915369 0
4915368 0
4915367 0
4915366 0
4915365 0
4915364 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



4857287 0
4857286 0
4857285 0
4857284 0
4857283 0
4857282 0
4857281 0
4857280 0
4857279 0
4857278 0
4857277 0
4857276 0
4857275 0
4857274 0
4857273 0
4857272 0
4857271 0
4857270 0
4857269 0
4857268 0
4857267 0
4857266 0
4857265 0
4857264 0
4857263 0
4857262 0
4857261 0
4857260 0
4857259 0
4857258 0
4857257 0
4857256 0
4857255 0
4857254 0
4857253 0
4857252 0
4857251 0
4857250 0
4857249 0
4857248 0
4857247 0
4857246 0
4857245 0
4857244 0
4857243 0
4857242 0
4857241 0
4857240 0
4857239 0
4857238 0
4857237 0
4857236 0
4857235 0
4857234 0
4857233 0
4857232 0
4857231 0
4857230 0
4857229 0
4857228 0
4857227 0
4857226 0
4857225 0
4857224 0
4857223 0
4857222 0
4857221 0
4857220 0
4857219 0
4857218 0
4857217 0
4857216 0
4857215 0
4857214 0
4857213 0
4857212 0
4857211 0
4857210 0
4857209 0
4857208 0
4857207 0
4857206 0
4857205 0
4857204 0
4857203 0
4857202 0
4857201 0
4857200 0
4857199 0
4857198 0
4857197 0
4857196 0
4857195 0
4857194 0
4857193 0
4857192 0
4857191 0
4857190 0
4857189 0
4857188 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



4772778 0
4772777 0
4772776 0
4772775 0
4772774 0
4772773 0
4772772 0
4772771 0
4772770 0
4772769 0
4772768 0
4772767 0
4772766 0
4772765 0
4772764 0
4772763 0
4772762 0
4772761 0
4772760 0
4772759 0
4772758 0
4772757 0
4772756 0
4772755 0
4772754 0
4772753 0
4772752 0
4772751 0
4772750 0
4772749 0
4772748 0
4772747 0
4772746 0
4772745 0
4772744 0
4772743 0
4772742 0
4772741 0
4772740 0
4772739 0
4772738 0
4772737 0
4772736 0
4772735 0
4772734 0
4772733 0
4772732 0
4772731 0
4772730 0
4772729 0
4772728 0
4772727 0
4772726 0
4772725 0
4772724 0
4772723 0
4772722 0
4772721 0
4772720 0
4772719 0
4772718 0
4772717 0
4772716 0
4772715 0
4772714 0
4772713 0
4772712 0
4772711 0
4772710 0
4772709 0
4772708 0
4772707 0
4772706 0
4772705 0
4772704 0
4772703 0
4772702 0
4772701 0
4772700 0
4772699 0
4772698 0
4772697 0
4772696 0
4772695 0
4772694 0
4772693 0
4772692 0
4772691 0
4772690 0
4772689 0
4772688 0
4772687 0
4772686 0
4772685 0
4772684 0
4772683 0
4772682 0
4772681 0
4772680 0
4772679 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



4640972 0
4640971 0
4640970 0
4640969 0
4640968 0
4640967 0
4640966 0
4640965 0
4640964 0
4640963 0
4640962 0
4640961 0
4640960 0
4640959 0
4640958 0
4640957 0
4640956 0
4640955 0
4640954 0
4640953 0
4640952 0
4640951 0
4640950 0
4640949 0
4640948 0
4640947 0
4640946 0
4640945 0
4640944 0
4640943 0
4640942 0
4640941 0
4640940 0
4640939 0
4640938 0
4640937 0
4640936 0
4640935 0
4640934 0
4640933 0
4640932 0
4640931 0
4640930 0
4640929 0
4640928 0
4640927 0
4640926 0
4640925 0
4640924 0
4640923 0
4640922 0
4640921 0
4640920 0
4640919 0
4640918 0
4640917 0
4640916 0
4640915 0
4640914 0
4640913 0
4640912 0
4640911 0
4640910 0
4640909 0
4640908 0
4640907 0
4640906 0
4640905 0
4640904 0
4640903 0
4640902 0
4640901 0
4640900 0
4640899 0
4640898 0
4640897 0
4640896 0
4640895 0
4640894 0
4640893 0
4640892 0
4640891 0
4640890 0
4640889 0
4640888 0
4640887 0
4640886 0
4640885 0
4640884 0
4640883 0
4640882 0
4640881 0
4640880 0
4640879 0
4640878 0
4640877 0
4640876 0
4640875 0
4640874 0
4640873 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



4527696 0
4527695 0
4527694 0
4527693 0
4527692 0
4527691 0
4527690 0
4527689 0
4527688 0
4527687 0
4527686 0
4527685 0
4527684 0
4527683 0
4527682 0
4527681 0
4527680 0
4527679 0
4527678 0
4527677 0
4527676 0
4527675 0
4527674 0
4527673 0
4527672 0
4527671 0
4527670 0
4527669 0
4527668 0
4527667 0
4527666 0
4527665 0
4527664 0
4527663 0
4527662 0
4527661 0
4527660 0
4527659 0
4527658 0
4527657 0
4527656 0
4527655 0
4527654 0
4527653 0
4527652 0
4527651 0
4527650 0
4527649 0
4527648 0
4527647 0
4527646 0
4527645 0
4527644 0
4527643 0
4527642 0
4527641 0
4527640 0
4527639 0
4527638 0
4527637 0
4527636 0
4527635 0
4527634 0
4527633 0
4527632 0
4527631 0
4527630 0
4527629 0
4527628 0
4527627 0
4527626 0
4527625 0
4527624 0
4527623 0
4527622 0
4527621 0
4527620 0
4527619 0
4527618 0
4527617 0
4527616 0
4527615 0
4527614 0
4527613 0
4527612 0
4527611 0
4527610 0
4527609 0
4527608 0
4527607 0
4527606 0
4527605 0
4527604 0
4527603 0
4527602 0
4527601 0
4527600 0
4527599 0
4527598 0
4527597 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



4459725 0
4459724 0
4459723 0
4459722 0
4459721 0
4459720 0
4459719 0
4459718 0
4459717 0
4459716 0
4459715 0
4459714 0
4459713 0
4459712 0
4459711 0
4459710 0
4459709 0
4459708 0
4459707 0
4459706 0
4459705 0
4459704 0
4459703 0
4459702 0
4459701 0
4459700 0
4459699 0
4459698 0
4459697 0
4459696 0
4459695 0
4459694 0
4459693 0
4459692 0
4459691 0
4459690 0
4459689 0
4459688 0
4459687 0
4459686 0
4459685 0
4459684 0
4459683 0
4459682 0
4459681 0
4459680 0
4459679 0
4459678 0
4459677 0
4459676 0
4459675 0
4459674 0
4459673 0
4459672 0
4459671 0
4459670 0
4459669 0
4459668 0
4459667 0
4459666 0
4459665 0
4459664 0
4459663 0
4459662 0
4459661 0
4459660 0
4459659 0
4459658 0
4459657 0
4459656 0
4459655 0
4459654 0
4459653 0
4459652 0
4459651 0
4459650 0
4459649 0
4459648 0
4459647 0
4459646 0
4459645 0
4459644 0
4459643 0
4459642 0
4459641 0
4459640 0
4459639 0
4459638 0
4459637 0
4459636 0
4459635 0
4459634 0
4459633 0
4459632 0
4459631 0
4459630 0
4459629 0
4459628 0
4459627 0
4459626 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



4394901 0
4394900 0
4394899 0
4394898 0
4394897 0
4394896 0
4394895 0
4394894 0
4394893 0
4394892 0
4394891 0
4394890 0
4394889 0
4394888 0
4394887 0
4394886 0
4394885 0
4394884 0
4394883 0
4394882 0
4394881 0
4394880 0
4394879 0
4394878 0
4394877 0
4394876 0
4394875 0
4394874 0
4394873 0
4394872 0
4394871 0
4394870 0
4394869 0
4394868 0
4394867 0
4394866 0
4394865 0
4394864 0
4394863 0
4394862 0
4394861 0
4394860 0
4394859 0
4394858 0
4394857 0
4394856 0
4394855 0
4394854 0
4394853 0
4394852 0
4394851 0
4394850 0
4394849 0
4394848 0
4394847 0
4394846 0
4394845 0
4394844 0
4394843 0
4394842 0
4394841 0
4394840 0
4394839 0
4394838 0
4394837 0
4394836 0
4394835 0
4394834 0
4394833 0
4394832 0
4394831 0
4394830 0
4394829 0
4394828 0
4394827 0
4394826 0
4394825 0
4394824 0
4394823 0
4394822 0
4394821 0
4394820 0
4394819 0
4394818 0
4394817 0
4394816 0
4394815 0
4394814 0
4394813 0
4394812 0
4394811 0
4394810 0
4394809 0
4394808 0
4394807 0
4394806 0
4394805 0
4394804 0
4394803 0
4394802 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



4305445 0
4305444 0
4305443 0
4305442 0
4305441 0
4305440 0
4305439 0
4305438 0
4305437 0
4305436 0
4305435 0
4305434 0
4305433 0
4305432 0
4305431 0
4305430 0
4305429 0
4305428 0
4305427 0
4305426 0
4305425 0
4305424 0
4305423 0
4305422 0
4305421 0
4305420 0
4305419 0
4305418 0
4305417 0
4305416 0
4305415 0
4305414 0
4305413 0
4305412 0
4305411 0
4305410 0
4305409 0
4305408 0
4305407 0
4305406 0
4305405 0
4305404 0
4305403 0
4305402 0
4305401 0
4305400 0
4305399 0
4305398 0
4305397 0
4305396 0
4305395 0
4305394 0
4305393 0
4305392 0
4305391 0
4305390 0
4305389 0
4305388 0
4305387 0
4305386 0
4305385 0
4305384 0
4305383 0
4305382 0
4305381 0
4305380 0
4305379 0
4305378 0
4305377 0
4305376 0
4305375 0
4305374 0
4305373 0
4305372 0
4305371 0
4305370 0
4305369 0
4305368 0
4305367 0
4305366 0
4305365 0
4305364 0
4305363 0
4305362 0
4305361 0
4305360 0
4305359 0
4305358 0
4305357 0
4305356 0
4305355 0
4305354 0
4305353 0
4305352 0
4305351 0
4305350 0
4305349 0
4305348 0
4305347 0
4305346 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



4004180 0
4004179 0
4004178 0
4004177 0
4004176 0
4004175 0
4004174 0
4004173 0
4004172 0
4004171 0
4004170 0
4004169 0
4004168 0
4004167 0
4004166 0
4004165 0
4004164 0
4004163 0
4004162 0
4004161 0
4004160 0
4004159 0
4004158 0
4004157 0
4004156 0
4004155 0
4004154 0
4004153 0
4004152 0
4004151 0
4004150 0
4004149 0
4004148 0
4004147 0
4004146 0
4004145 0
4004144 0
4004143 0
4004142 0
4004141 0
4004140 0
4004139 0
4004138 0
4004137 0
4004136 0
4004135 0
4004134 0
4004133 0
4004132 0
4004131 0
4004130 0
4004129 0
4004128 0
4004127 0
4004126 0
4004125 0
4004124 0
4004123 0
4004122 0
4004121 0
4004120 0
4004119 0
4004118 0
4004117 0
4004116 0
4004115 0
4004114 0
4004113 0
4004112 0
4004111 0
4004110 0
4004109 0
4004108 0
4004107 0
4004106 0
4004105 0
4004104 0
4004103 0
4004102 0
4004101 0
4004100 0
4004099 0
4004098 0
4004097 0
4004096 0
4004095 0
4004094 0
4004093 0
4004092 0
4004091 0
4004090 0
4004089 0
4004088 0
4004087 0
4004086 0
4004085 0
4004084 0
4004083 0
4004082 0
4004081 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



3901864 0
3901863 0
3901862 0
3901861 0
3901860 0
3901859 0
3901858 0
3901857 0
3901856 0
3901855 0
3901854 0
3901853 0
3901852 0
3901851 0
3901850 0
3901849 0
3901848 0
3901847 0
3901846 0
3901845 0
3901844 0
3901843 0
3901842 0
3901841 0
3901840 0
3901839 0
3901838 0
3901837 0
3901836 0
3901835 0
3901834 0
3901833 0
3901832 0
3901831 0
3901830 0
3901829 0
3901828 0
3901827 0
3901826 0
3901825 0
3901824 0
3901823 0
3901822 0
3901821 0
3901820 0
3901819 0
3901818 0
3901817 0
3901816 0
3901815 0
3901814 0
3901813 0
3901812 0
3901811 0
3901810 0
3901809 0
3901808 0
3901807 0
3901806 0
3901805 0
3901804 0
3901803 0
3901802 0
3901801 0
3901800 0
3901799 0
3901798 0
3901797 0
3901796 0
3901795 0
3901794 0
3901793 0
3901792 0
3901791 0
3901790 0
3901789 0
3901788 0
3901787 0
3901786 0
3901785 0
3901784 0
3901783 0
3901782 0
3901781 0
3901780 0
3901779 0
3901778 0
3901777 0
3901776 0
3901775 0
3901774 0
3901773 0
3901772 0
3901771 0
3901770 0
3901769 0
3901768 0
3901767 0
3901766 0
3901765 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



3751059 0
3751058 0
3751057 0
3751056 0
3751055 0
3751054 0
3751053 0
3751052 0
3751051 0
3751050 0
3751049 0
3751048 0
3751047 0
3751046 0
3751045 0
3751044 0
3751043 0
3751042 0
3751041 0
3751040 0
3751039 0
3751038 0
3751037 0
3751036 0
3751035 0
3751034 0
3751033 0
3751032 0
3751031 0
3751030 0
3751029 0
3751028 0
3751027 0
3751026 0
3751025 0
3751024 0
3751023 0
3751022 0
3751021 0
3751020 0
3751019 0
3751018 0
3751017 0
3751016 0
3751015 0
3751014 0
3751013 0
3751012 0
3751011 0
3751010 0
3751009 0
3751008 0
3751007 0
3751006 0
3751005 0
3751004 0
3751003 0
3751002 0
3751001 0
3751000 0
3750999 0
3750998 0
3750997 0
3750996 0
3750995 0
3750994 0
3750993 0
3750992 0
3750991 0
3750990 0
3750989 0
3750988 0
3750987 0
3750986 0
3750985 0
3750984 0
3750983 0
3750982 0
3750981 0
3750980 0
3750979 0
3750978 0
3750977 0
3750976 0
3750975 0
3750974 0
3750973 0
3750972 0
3750971 0
3750970 0
3750969 0
3750968 0
3750967 0
3750966 0
3750965 0
3750964 0
3750963 0
3750962 0
3750961 0
3750960 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



3659286 0
3659285 0
3659284 0
3659283 0
3659282 0
3659281 0
3659280 0
3659279 0
3659278 0
3659277 0
3659276 0
3659275 0
3659274 0
3659273 0
3659272 0
3659271 0
3659270 0
3659269 0
3659268 0
3659267 0
3659266 0
3659265 0
3659264 0
3659263 0
3659262 0
3659261 0
3659260 0
3659259 0
3659258 0
3659257 0
3659256 0
3659255 0
3659254 0
3659253 0
3659252 0
3659251 0
3659250 0
3659249 0
3659248 0
3659247 0
3659246 0
3659245 0
3659244 0
3659243 0
3659242 0
3659241 0
3659240 0
3659239 0
3659238 0
3659237 0
3659236 0
3659235 0
3659234 0
3659233 0
3659232 0
3659231 0
3659230 0
3659229 0
3659228 0
3659227 0
3659226 0
3659225 0
3659224 0
3659223 0
3659222 0
3659221 0
3659220 0
3659219 0
3659218 0
3659217 0
3659216 0
3659215 0
3659214 0
3659213 0
3659212 0
3659211 0
3659210 0
3659209 0
3659208 0
3659207 0
3659206 0
3659205 0
3659204 0
3659203 0
3659202 0
3659201 0
3659200 0
3659199 0
3659198 0
3659197 0
3659196 0
3659195 0
3659194 0
3659193 0
3659192 0
3659191 0
3659190 0
3659189 0
3659188 0
3659187 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



3469783 0
3469782 0
3469781 0
3469780 0
3469779 0
3469778 0
3469777 0
3469776 0
3469775 0
3469774 0
3469773 0
3469772 0
3469771 0
3469770 0
3469769 0
3469768 0
3469767 0
3469766 0
3469765 0
3469764 0
3469763 0
3469762 0
3469761 0
3469760 0
3469759 0
3469758 0
3469757 0
3469756 0
3469755 0
3469754 0
3469753 0
3469752 0
3469751 0
3469750 0
3469749 0
3469748 0
3469747 0
3469746 0
3469745 0
3469744 0
3469743 0
3469742 0
3469741 0
3469740 0
3469739 0
3469738 0
3469737 0
3469736 0
3469735 0
3469734 0
3469733 0
3469732 0
3469731 0
3469730 0
3469729 0
3469728 0
3469727 0
3469726 0
3469725 0
3469724 0
3469723 0
3469722 0
3469721 0
3469720 0
3469719 0
3469718 0
3469717 0
3469716 0
3469715 0
3469714 0
3469713 0
3469712 0
3469711 0
3469710 0
3469709 0
3469708 0
3469707 0
3469706 0
3469705 0
3469704 0
3469703 0
3469702 0
3469701 0
3469700 0
3469699 0
3469698 0
3469697 0
3469696 0
3469695 0
3469694 0
3469693 0
3469692 0
3469691 0
3469690 0
3469689 0
3469688 0
3469687 0
3469686 0
3469685 0
3469684 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



3398394 0
3398393 0
3398392 0
3398391 0
3398390 0
3398389 0
3398388 0
3398387 0
3398386 0
3398385 0
3398384 0
3398383 0
3398382 0
3398381 0
3398380 0
3398379 0
3398378 0
3398377 0
3398376 0
3398375 0
3398374 0
3398373 0
3398372 0
3398371 0
3398370 0
3398369 0
3398368 0
3398367 0
3398366 0
3398365 0
3398364 0
3398363 0
3398362 0
3398361 0
3398360 0
3398359 0
3398358 0
3398357 0
3398356 0
3398355 0
3398354 0
3398353 0
3398352 0
3398351 0
3398350 0
3398349 0
3398348 0
3398347 0
3398346 0
3398345 0
3398344 0
3398343 0
3398342 0
3398341 0
3398340 0
3398339 0
3398338 0
3398337 0
3398336 0
3398335 0
3398334 0
3398333 0
3398332 0
3398331 0
3398330 0
3398329 0
3398328 0
3398327 0
3398326 0
3398325 0
3398324 0
3398323 0
3398322 0
3398321 0
3398320 0
3398319 0
3398318 0
3398317 0
3398316 0
3398315 0
3398314 0
3398313 0
3398312 0
3398311 0
3398310 0
3398309 0
3398308 0
3398307 0
3398306 0
3398305 0
3398304 0
3398303 0
3398302 0
3398301 0
3398300 0
3398299 0
3398298 0
3398297 0
3398296 0
3398295 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



3268063 0
3268062 0
3268061 0
3268060 0
3268059 0
3268058 0
3268057 0
3268056 0
3268055 0
3268054 0
3268053 0
3268052 0
3268051 0
3268050 0
3268049 0
3268048 0
3268047 0
3268046 0
3268045 0
3268044 0
3268043 0
3268042 0
3268041 0
3268040 0
3268039 0
3268038 0
3268037 0
3268036 0
3268035 0
3268034 0
3268033 0
3268032 0
3268031 0
3268030 0
3268029 0
3268028 0
3268027 0
3268026 0
3268025 0
3268024 0
3268023 0
3268022 0
3268021 0
3268020 0
3268019 0
3268018 0
3268017 0
3268016 0
3268015 0
3268014 0
3268013 0
3268012 0
3268011 0
3268010 0
3268009 0
3268008 0
3268007 0
3268006 0
3268005 0
3268004 0
3268003 0
3268002 0
3268001 0
3268000 0
3267999 0
3267998 0
3267997 0
3267996 0
3267995 0
3267994 0
3267993 0
3267992 0
3267991 0
3267990 0
3267989 0
3267988 0
3267987 0
3267986 0
3267985 0
3267984 0
3267983 0
3267982 0
3267981 0
3267980 0
3267979 0
3267978 0
3267977 0
3267976 0
3267975 0
3267974 0
3267973 0
3267972 0
3267971 0
3267970 0
3267969 0
3267968 0
3267967 0
3267966 0
3267965 0
3267964 0


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



2790438 0
2790437 0
2790436 0
2790435 0
2790434 0
2790433 0
2790432 0
2790431 0
2790430 0
2790429 0
2790428 0
2790427 0
2790426 0
2790425 0
2790424 0
2790423 0
2790422 0
2790421 0
2790420 0
2790419 0
2790418 0
2790417 0
2790416 0
2790415 0
2790414 0
2790413 0
2790412 0
2790411 0
2790410 0
2790409 0
2790408 0
2790407 0
2790406 0
2790405 0
2790404 0
2790403 0
2790402 0
2790401 0
2790400 0
2790399 0
2790398 0
2790397 0
2790396 0
2790395 0
2790394 0
2790393 0
2790392 0
2790391 0
2790390 0
2790389 0
2790388 0
2790387 0
2790386 0
2790385 0
2790384 0
2790383 0
2790382 0
2790381 0
2790380 0
2790379 0
2790378 0
2790377 0
2790376 0
2790375 0
2790374 0
2790373 0
2790372 0
2790371 0
2790370 0
2790369 0
2790368 0
2790367 0
2790366 0
2790365 0
2790364 0
2790363 0
2790362 0
2790361 0
2790360 0
2790359 0
2790358 0
2790357 0
2790356 0
2790355 0
2790354 0
2790353 0
2790352 0
2790351 0
2790350 0
2790349 0
2790348 0
2790347 0
2790346 0
2790345 0
2790344 0
2790343 0
2790342 0
2790341 0
2790340 0
2790339 0


In [62]:
open Big_int

error: compile_error

In [61]:
let factorial : big_int -> big_int = fun num ->
     let rec factTR : big_int -> big_int -> big_int = fun n acc ->
          if gt_big_int n (big_int_of_int 0)
          then factTR 
            (sub_big_int n (big_int_of_int 1)) 
            (mult_big_int acc n)
          else acc
     in
          factTR num (big_int_of_int 1)

error: compile_error